File: src/main/java/io/j99/idea/vue/parser/VueParserDefinition.java
package io.j99.idea.vue.parser;

import com.intellij.lang.ASTNode;
import com.intellij.lang.ParserDefinition;
import com.intellij.lang.PsiParser;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.project.Project;
import com.intellij.psi.FileViewProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.tree.IFileElementType;
import com.intellij.psi.tree.TokenSet;
import io.j99.idea.vue.VueFileType;
import io.j99.idea.vue.VueTokenizer;
import io.j99.idea.vue.lexer.VueLexer2;
import io.j99.idea.vue.lexer.VueTokenTypes;
import io.j99.idea.vue.psi.VueFile;
import io.j99.idea.vue.psi.impl.TemplateBodyCodeImpl;
import io.j99.idea.vue.psi.impl.VueElementImpl;
import org.jetbrains.annotations.NotNull;

/**
 * Created by android on 1/14/16.
 */
public class VueParserDefinition implements ParserDefinition {
    private static final IFileElementType FILE_ELEMENT_TYPE = new IFileElementType(VueFileType.INSTANCE.getLanguage());

    @NotNull
    @Override
    public Lexer createLexer(Project project) {
        return new VueLexer2(new VueTokenizer());
    }

    @Override
    public PsiParser createParser(Project project) {
        return new VueParser();
    }

    @Override
    public IFileElementType getFileNodeType() {
        return FILE_ELEMENT_TYPE;
    }

    @NotNull
    @Override
    public TokenSet getWhitespaceTokens() {
        return TokenSet.create(VueTokenTypes.WHITE_SPACE);
    }

    @NotNull
    @Override
    public TokenSet getCommentTokens() {
        return  TokenSet.create();
    }

    @NotNull
    @Override
    public TokenSet getStringLiteralElements() {
        return TokenSet.create();
    }

    @NotNull
    @Override
    public PsiElement createElement(ASTNode node) {
        if(node.getElementType()==VueTokenTypes.TAG_SCRIPT_BODY){
            return new TemplateBodyCodeImpl(node);
        }
        return new VueElementImpl(node);
    }

    @Override
    public PsiFile createFile(FileViewProvider fileViewProvider) {
        return new VueFile(fileViewProvider);
    }

    @Override
    public SpaceRequirements spaceExistanceTypeBetweenTokens(ASTNode astNode, ASTNode astNode1) {
        return SpaceRequirements.MAY;
    }
}

File: src/main/kotlin/io/j99/idea/vue/VueParserDefinition.kt
package io.j99.idea.vue

import com.intellij.lang.ASTNode
import com.intellij.lexer.FlexAdapter
import com.intellij.lexer.Lexer
import com.intellij.openapi.project.Project
import com.intellij.psi.FileViewProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import io.j99.idea.vue.psi.VueFile

/**
 * Created by android on 15-12-25.
 */
class VueParserDefinition : com.intellij.lang.xhtml.XHTMLParserDefinition() {
    override fun createFile(fileViewProvider: FileViewProvider): PsiFile {
        return VueFile(fileViewProvider)
    }

    override fun createLexer(p0: Project?): Lexer {
//        return FlexAdapter(_VueLexter())
        return super.createLexer(p0)
    }

    override fun createElement(p0: ASTNode?): PsiElement {
        return super.createElement(p0)
//        return VueTypes.Factory.createElement(p0);
    }
}
